contracts/vault/VaultBase.sol

```diff
@@ -0,0 +1,288 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.28;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "./Accounts.sol";
+import "./Locks.sol";
+
+/// Records account balances and token flows. Accounts are separated into funds.
+/// Funds are kept separate between controllers.
+///
+/// A fund can only be manipulated by a controller when it is locked. Tokens can
+/// only be withdrawn when a fund is unlocked.
+///
+/// The vault maintains a number of invariants to ensure its integrity.
+///
+/// The lock invariant ensures that there is a maximum time that a fund can be
+/// locked:
+///
+/// (∀ controller ∈ Controller, fund ∈ Fund:
+///   lock.expiry <= lock.maximum
+///   where lock = _locks[controller][fund])
+///
+/// The account invariant ensures that the outgoing token flow can be sustained
+/// for the maximum time that a fund can be locked:
+///
+/// (∀ controller ∈ Controller, fund ∈ Fund, account ∈ AccountId:
+///   flow.outgoing * (lock.maximum - flow.updated) <= balance.available
+///   where lock = _locks[controller][fund])
+///   and flow = _accounts[controller][fund][account].flow
+///   and balance = _accounts[controller][fund][account].balance
+///
+/// The flow invariant ensures that incoming and outgoing flow rates match:
+///
+/// (∀ controller ∈ Controller, fund ∈ Fund:
+///   (∑ account ∈ AccountId: accounts[account].flow.incoming) =
+///   (∑ account ∈ AccountId: accounts[account].flow.outgoing)
+///   where accounts = _accounts[controller][fund])
+///
+abstract contract VaultBase {
+  using SafeERC20 for IERC20;
+  using Accounts for Account;
+  using Locks for Lock;
+
+  IERC20 internal immutable _token;
+
+  /// Represents a smart contract that can redistribute and burn tokens in funds
+  type Controller is address;
+  /// Unique identifier for a fund, chosen by the controller
+  type Fund is bytes32;
+
+  /// Each fund has its own time lock
+  mapping(Controller => mapping(Fund => Lock)) private _locks;
+  /// Each account holder has its own set of accounts in a fund
+  mapping(Controller => mapping(Fund => mapping(AccountId => Account)))
+    private _accounts;
+
+  constructor(IERC20 token) {
+    _token = token;
+  }
+
+  function _getLockStatus(
+    Controller controller,
+    Fund fund
+  ) internal view returns (LockStatus) {
+    return _locks[controller][fund].status();
+  }
+
+  function _getLockExpiry(
+    Controller controller,
+    Fund fund
+  ) internal view returns (Timestamp) {
+    return _locks[controller][fund].expiry;
+  }
+
+  function _getBalance(
+    Controller controller,
+    Fund fund,
+    AccountId id
+  ) internal view returns (Balance memory) {
+    Lock memory lock = _locks[controller][fund];
+    LockStatus lockStatus = lock.status();
+    if (lockStatus == LockStatus.Locked) {
+      Account memory account = _accounts[controller][fund][id];
+      account.update(Timestamps.currentTime());
```

Comment:
I find `update` a bit awkward in the context of a `view`, where nothing is being stored, except for in memory. Maybe `atEnd` or `simulateEndAt`...?

Reply:
I also find it bit hard to find a better name, because it updates the struct that you pass as a parameter. Names such as 'atEnd()` or `simulateAtEnd()` suggest to me that there would be a return value with the calculated values.

Maybe `updateFlow()`?

Reply:
Maybe `updateSelf`? Or `updateSelfAtEnd`?